text/0000-async-stream.md

```diff
@@ -0,0 +1,754 @@
+returning [`Poll::pending`] implies that the stream has arranged for
+the current task to be re-awoken when the data is ready.
+
+[iterator]: https://doc.rust-lang.org/std/iter/trait.Iterator.html
+[`Future`]: https://doc.rust-lang.org/std/future/trait.Future.html
+[`Poll::pending`]: https://doc.rust-lang.org/std/task/enum.Poll.html#variant.Pending
+
+```rust
+// Defined in std::stream module
+pub trait Stream {
+    // Core items:
+    type Item;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+    
+    // Optional optimization hint, just like with iterators:
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+
+    // Convenience methods (covered later on in the RFC):
+    fn next(&mut self) -> Next<'_, Self>
+    where
+        Self: Unpin;
+}
+```
+
+The arguments to `poll_next` match that of the [`Future::poll`] method:
+
+* The self must be a pinned reference, ensuring both unique access to
+  the stream and that the stream value itself will not move. Pinning
+  allows the stream to save pointers into itself when it suspends,
+  which will be required to support generator syntax at some point.
+* The [context] `cx` defines details of the current task. In particular,
+  it gives access to the [`Waker`] for the task, which will allow the
+  task to be re-awoken once data is ready.
+
+[`Future::poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
+[pinned]: https://doc.rust-lang.org/std/pin/struct.Pin.html
+[context]: https://doc.rust-lang.org/std/task/struct.Context.html
+[`Waker`]: https://doc.rust-lang.org/std/task/struct.Waker.html
+
+### Why does next require Self:Unpin?
+
+When drafting this RFC, there was a [good deal of discussion](https://github.com/rust-lang/wg-async-foundations/pull/15#discussion_r452482084) around why the `next` method requires `Self:Unpin`.
+
+To understand this, it helps to take a closer look at the definition of `Next` (this struct is further discussed later in this RFC) in the [futures-util crate](https://docs.rs/futures-util/0.3.5/src/futures_util/stream/stream/next.rs.html#10-12).
+
+```rust
+pub struct Next<'a, St: ?Sized> {
+    stream: &'a mut St,
+}
+```
+Since `Stream::poll_next` takes a pinned reference, the next future needs `S` to be `Unpin` in order to safely construct a `Pin<&mut S>` from a `&mut S`.
+
+An alternative approach we could take would be to have the `next` method take `Pin<&mut S>`, rather than `&mut S`. However, this would require pinning even when the type is `Unpin`. The current approach requires pinning only when the type is not `Unpin`.
+
+At the moment, we do not see many `!Unpin` streams in practice (though there is one in the [futures-intrusive crate](https://github.com/Matthias247/futures-intrusive/blob/master/src/channel/mpmc.rs#L565-L625)). Where they will become important is when we introduce async generators, as discussed in [future-possibilities].
```

Comment:
I see them not uncommonly with `stream.then(|_| async {})` resulting in a `!Unpin` stream because of the future.

Reply:
Added that in as a clarification. Good catch!